### PR TITLE
feat: implement online style fetching for stable style resolving

### LINKED ISSUE #439: Implement style.json resolution strategy a when handling the stable style url
## Description

A frontend client is provided a single endpoint to use for retrieving a style.json definition (see #438). This endpoint should be handled by the backend, which performs the following resolution strategy to determine the style.json payload that's returned as the response:

1. Check if there's a offline static map with the style id of `default` (or whatever name we decide on). In practice, this means that there's a file that exists on the filesystem at `__ROOT__/default/style.json`, where `__ROOT__` is a provided directory where static custom maps are stored. If it exists, serve the style.json content as the response

2. if internet access is available, proxy a default online style of our choosing (e.g. using a mapbox style that requires our account's public access token)

3. if internet access is not available, serve an fallback map style that uses statically served local geojson files (e.g. https://github.com/digidem/mapeo-offline-map) 

## Tasks

- [ ] Add route handler for the stable style.json url implemented in #438, which does the described resolution strategy

### LINKED ISSUE #448: Implement style.json resolution strategy a when handling the stable style url #2
linked to [Implement style.json resolution strategy a when handling the stable style url #439](https://github.com/digidem/mapeo-core-next/issues/439)

Description
A frontend client is provided a single endpoint to use for retrieving a style.json definition (see https://github.com/digidem/mapeo-core-next/issues/438). This endpoint should be handled by the backend, which performs the following resolution strategy to determine the style.json payload that's returned as the response:

1-Check if there's a offline static map with the style id of default (or whatever name we decide on). In practice, this means that there's a file that exists on the filesystem at __ROOT__/default/style.json, where __ROOT__ is a provided directory where static custom maps are stored. If it exists, serve the style.json content as the response

2- if internet access is available, proxy a default online style of our choosing (e.g. using a mapbox style that requires our account's public access token)
3- if internet access is not available, serve an fallback map style that uses statically served local geojson files (e.g. https://github.com/digidem/mapeo-offline-map)

##Todo
- if internet access is available, proxy a default online style of our choosing (e.g. using a mapbox style that requires our account's public access token)

### PR DESCRIPTION
**Pull request by achou11**
Closes #439 
Closes #448

Stacked on #450 

Notes:

- implements the online proxying component of determining which style.json to return handling the stable style url endpoint. Currently assumes that the client will provide an `key` query param when requesting the stable style json endpoint (as opposed to us using some hardcoded token)
- e2e test(s) will be implemented once #457 is merged

TODO:

- [X] ~Potentially use a proper fixture and/or making an actual request to the upstream provider? Right now I mock the response using Undici's mocking mechanisms~ Added fixtures for style.json's from both Mapbox and Protomaps 
- ~Do we want to incorporate a compression plugin? (e.g. https://github.com/fastify/fastify-compress/). Most of the time, these upstream providers are compressing the responses but since we don't have anything set up, our forwarded response is not compressed right now~ EDIT: seems like we don't need to right now based on feedback

### CODE REVIEW

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [inline src/fastify-plugins/maps/index.js:85]: I don't see this as a TODO. I think we want to use our own compression regardless of what the upstream source says.

```suggestion
              // We do our own content encoding.
```

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [reply src/fastify-plugins/maps/index.js:79]: Should this have a timeout?

> should i forward the `req.headers` here? or maybe specify the `credential` opt?

I wouldn't until it becomes a problem. I think it's hard to anticipate our needs.

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [reply src/fastify-plugins/maps/index.js:15]: I don't know, but I think it's probably fine to merge this as long as we eventually confirm this is reasonable.

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [reply src/fastify-plugins/maps/index.js:86]: I don't see much value in compressing here because this is running locally. I suspect the overhead of compression will probably be larger than the overhead of IPC data transfer. We'd have to profile that, of course.

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [inline src/fastify-plugins/maps/index.js:102]: 1. I think this should log a warning, not an error. No developer has made a mistake if this request fails.
2. Should we add a little info about what went wrong?

```suggestion
          fastify.log.error('Upstream style.json request totally failed:', err)
```

File: tests/fastify-plugins/maps.js
Comment (EvanHahn) [inline tests/fastify-plugins/maps.js:216]: Do we ever undo this global dispatcher?

File: tests/fastify-plugins/maps.js
Comment (EvanHahn) [reply tests/fastify-plugins/maps.js:208]: Probably not worth it, but I could see creating this `MockAgent` and passing it into the `fetch` using [the `dispatcher` option](https://undici.nodejs.org/#/?id=undicifetchinput-init-promise). I think what you have here is fine.

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [reply src/fastify-plugins/maps/index.js:168]: I don't know. Perhaps, though, we could add a warning if we see a hostname we don't expect?

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [reply src/fastify-plugins/maps/index.js:166]: I don't know much about mapping APIs but I don't think we should assume anything about the upstream provider. Personally, I think it's fine to remove this TODO and leave things as is.

**Review by EvanHahn — APPROVED** [review]
LGTM other than my numerous nitpicks.

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js:79]: > Should this have a timeout?

probably yeah. thoughts on a reasonable number?

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js:83]: yeah that make sense. not sure about which headers are essential though (there are quite a few from these providers 😄 ). happy to change if we can narrow down the ones we want to actually forward

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js:85]: I still consider it a TODO because there's no content-encoding implemented on our end yet. But yeah mostly added. the note to remind us that we should probably have something implemented

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js:102]: initial thought is that it's a little too similar in wording to the nested message about getting a bad response. maybe one of these should be reworded to have more differentiation?

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js:128]: ah yeah i had initially called it `apiKey` but renamed the variables to `key`. will update these docs (and add some for the parameter itself)

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js:161]: nope a bad copy-paste. thanks for catching!

File: test-e2e/manager-fastify-server.js
Comment (achou11) [reply test-e2e/manager-fastify-server.js:286]: nope - a remnant of about 3 refactors 😄

File: tests/fastify-plugins/maps.js
Comment (achou11) [reply tests/fastify-plugins/maps.js:216]: in its current implementation, no. was wondering if i needed to and had a previous implementation where i basically called this setup for each test so that I could access the teardown. maybe that's the safe way to go?

i think part of it depends on how the test runner works. does brittle run all tests in the same process? if so, will probably need some kind of cleanup

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js:83]: @gmaclennan any thoughts?

File: tests/fastify-plugins/maps.js
Comment (achou11) [reply tests/fastify-plugins/maps.js:208]: useful to know that, but I don't think it would handle the case where the code we're testing is calling fetch internally. the global dispatcher approach makes sure that we can intercept those underlying fetch calls

File: tests/fastify-plugins/maps.js
Comment (achou11) [reply tests/fastify-plugins/maps.js:216]: the hacky approach is to add a final `test` that does a cleanup e.g.


test('cleanup fetch mock', (t) => {
  t.teardown(() => {
    cleanupFetchMock()
  })
})

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [reply src/fastify-plugins/maps/index.js:79]: Not sure...30 seconds? 60 seconds?

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [reply src/fastify-plugins/maps/index.js:102]: Yeah...I think there's a "network error" case and a "non-2XX error". Maybe we update the log messages with that kind of thing?

File: tests/fastify-plugins/maps.js
Comment (EvanHahn) [reply tests/fastify-plugins/maps.js:216]: Alternatively, you could put everything in a `test` block and then have subtests. Something vaguely like:

```js
test('outer test', async (t) => {
  setupFetchMock()
  t.teardown(cleanupFetchMock)

  t.test('subtest 1', async (st) => { /* ... */ })

  t.test('subtest 2', async (st) => { /* ... */ })

  // ...
})
```

Comment (achou11) [reply tests/fastify-plugins/maps.js:216]: i'll leave it as is for now if that's okay then!

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/42ad04bb1d00202887116004bc7f1556e628e9c7

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/b94ccef8d0f5cb25f2089a1e3e40951e7b9542f7

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/05cb7c5f198257fa7bca97d395203e859d895cc8

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/b94ccef8d0f5cb25f2089a1e3e40951e7b9542f7

File: test-e2e/manager-fastify-server.js
Comment (achou11) [reply test-e2e/manager-fastify-server.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/0d56a6c6d7f8a929e4e52f67a40f948038338cd2

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/af0d737ace5f9a0895fc696a26b77d469d19001a

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/0654550d7c90f2cbaf3368421b97853324decf15

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/3c121ad090426aedcc8d11c368237e9fac5d645c

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/b94ccef8d0f5cb25f2089a1e3e40951e7b9542f7

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/b94ccef8d0f5cb25f2089a1e3e40951e7b9542f7

**Comment by EvanHahn** [comment]
I plan to review this on Monday.

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [reply src/fastify-plugins/maps/index.js]: TIL about `AbortSignal.timeout`.

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [reply src/fastify-plugins/maps/index.js:121]: Short answer: yes, I think it should be a warning.

---

Long answer:

At a high level, I believe two things about how errors should be logged:

1. _Errors are for developer mistakes._ Invalid states are a good example, because a perfect developer would never put the system into an invalid state.
1. _Warnings are for unavoidable problems._ Network failures are usually a good example, because even a perfect developer can't fix a user turning their wifi off.

To be clear, this is personal preference!

So what would I do this case?

If we get a 5xx error from the server, that's a _warning_, because we didn't make a mistake. But you could argue that we should _error_ if we get a 4xx. After all, if we send a bogus request, isn't that a developer's fault?

In this case, I think the answer is..._maybe_? Do we consider a bad API key an error, or a warning? Does it depend on what kind of 4xx status code you get (e.g., is 422 "our fault" but 429 is not)? I don't think it's worth thinking too hard about, so I'd probably just go with a warning.

Comment (achou11) [reply src/fastify-plugins/maps/index.js:121]: this is helpful! thanks for sharing both answers 😄

File: src/fastify-plugins/maps/index.js
Comment (EvanHahn) [inline src/fastify-plugins/maps/index.js:88]: Currently, the flow is:

1. Determine if the map provider is in a list with `isRecognizedMapProvider`.
1. If it is, pass it to `getMapProviderApiKeyQueryParamName` (which could theoretically throw an error) and replace the query string.
1. If it's not, log a warning.

I wonder if we could avoid error states completely by changing the flow to:

1. Look up `paramToUpsert` in a Map.
1. If it's found, replace it.
1. If it's not, log a warning.

This solution (which might be worse for some reason) is ~35 lines shorter:

```diff
diff --git a/src/fastify-plugins/maps/index.js b/src/fastify-plugins/maps/index.js
index 6fc4cfb..9e48e67 100644
--- a/src/fastify-plugins/maps/index.js
+++ b/src/fastify-plugins/maps/index.js
@@ -14,12 +14,17 @@ export const PLUGIN_NAME = 'mapeo-maps'
 export const DEFAULT_MAPBOX_STYLE_URL =
   'https://api.mapbox.com/styles/v1/mapbox/outdoors-v12'
 
-const RECOGNIZED_MAP_PROVIDERS = /** @type {const} */ ([
-  'api.mapbox.com',
-  'api.protomaps.com',
-  'api.maptiler.com',
-  'tiles.stadiamaps.com',
-  'basemapstyles-api.arcgis.com',
+const MAP_PROVIDER_API_KEY_QUERY_PARAM_BY_HOSTNAME = new Map([
+  // Mapbox expects `access_token`: https://docs.mapbox.com/api/maps/styles/
+  ['api.mapbox.com', 'access_token'],
+  // Protomaps expects `key` (no docs link yet)
+  ['api.protomaps.com', 'key'],
+  // MapTiler expects `key`: https://docs.maptiler.com/cloud/api/maps/
+  ['api.maptiler.com', 'key'],
+  // Stadia expects `api_key`: https://docs.stadiamaps.com/themes/
+  ['tiles.stadiamaps.com', 'api_key'],
+  // ArcGIS expects `token`: https://developers.arcgis.com/documentation/mapping-apis-and-services/security/api-keys/
+  ['basemapstyles-api.arcgis.com', 'token'],
 ])
 
 export const plugin = fp(mapsPlugin, {
@@ -84,9 +89,9 @@ async function routes(fastify, opts) {
         const { hostname } = upstreamUrlObj
 
         if (key) {
-          if (isRecognizedMapProvider(hostname)) {
-            const paramToUpsert = getMapProviderApiKeyQueryParamName(hostname)
-
+          const paramToUpsert =
+            MAP_PROVIDER_API_KEY_QUERY_PARAM_BY_HOSTNAME.get(hostname)
+          if (paramToUpsert) {
             // Note that even if the search param of interest already exists in the url
             // it is overwritten by the key provided in the request's search params
             upstreamUrlObj.searchParams.set(paramToUpsert, key)
@@ -147,43 +152,3 @@ async function routes(fastify, opts) {
     }
   )
 }
-
-/**
- * @param {string} hostname URL hostname
- * @returns {hostname is typeof RECOGNIZED_MAP_PROVIDERS[number]}
- */
-function isRecognizedMapProvider(hostname) {
-  return RECOGNIZED_MAP_PROVIDERS.includes(
-    /** @type {typeof RECOGNIZED_MAP_PROVIDERS[number]} */ (hostname)
-  )
-}
-
-/**
- * @param {typeof RECOGNIZED_MAP_PROVIDERS[number]} hostname
- * @returns {string} The name of the URL search param used by a provider to specify an auth-related API key or token
- */
-function getMapProviderApiKeyQueryParamName(hostname) {
-  switch (hostname) {
-    // Mapbox expects `access_token`: https://docs.mapbox.com/api/maps/styles/
-    case 'api.mapbox.com': {
-      return 'access_token'
-    }
-    // Protomaps expects `key` (no docs link yet)
-    // MapTiler expects `key`: https://docs.maptiler.com/cloud/api/maps/
-    case 'api.protomaps.com':
-    case 'api.maptiler.com': {
-      return 'key'
-    }
-    // Stadia expects `api_key`: https://docs.stadiamaps.com/themes/
-    case 'tiles.stadiamaps.com': {
-      return 'api_key'
-    }
-    // ArcGIS expects `token`: https://developers.arcgis.com/documentation/mapping-apis-and-services/security/api-keys/
-    case 'basemapstyles-api.arcgis.com': {
-      return 'token'
-    }
-    default: {
-      throw new Error(`Invalid value provided ${hostname}`)
-    }
-  }
-}
```

Take it or leave it!

**Review by EvanHahn — APPROVED** [review]
LGTM as long as we get an answer to this question (doesn't have to be before merge):

https://github.com/digidem/mapeo-core-next/pull/459/files#r1473479880

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js:88]: good suggestion! makes way more sense looking back on it :)

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/0d992efff1f0e0f13b17a89478bd4dbd32079fed

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js]: addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/c0c25d9fa5204f76a5d04c026da0b039368afe58

File: src/fastify-plugins/maps/index.js
Comment (achou11) [reply src/fastify-plugins/maps/index.js:113]: i updated this implementation in https://github.com/digidem/mapeo-core-next/pull/459/commits/ef1c0dd621781363835150d645c2dde83e8486b9 to use an allow-list approach and currently have it so that it's only forwarding headers related to caching.

**Review by EvanHahn — APPROVED** [review]

**Comment by EvanHahn** [comment]
Not planning to re-review this. Let me know if that's wrong!

**Comment by EvanHahn** [comment]
@achou11 Anything else you want to do on this PR before we merge it?

**Comment by achou11** [comment]
> @achou11 Anything else you want to do on this PR before we merge it?

considering it's stacked, I'd prefer the merge the other ones first based on stack order. the other ones haven't been reviewed yet and so id refrain from merging in order to make them easier to review

**Comment by EvanHahn** [comment]
Ah, of course. Sorry for the ping!

**Pull request merged**
